src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
+                peers.removeIf {
+                    it !in getTopicPeers(topic) || score.score(it) < scoreParams.publishThreshold
+                }
+                val needMore = coreParams.D - peers.size
                 if (needMore > 0) {
-                    peers += (getTopicPeers(topic) - peers).shuffled(random).take(needMore)
+                    peers += (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= scoreParams.publishThreshold && !isDirect(it) }
+                        .shuffled(random)
+                        .take(needMore)
                 }
-                submitGossip(topic, peers)
+                emitGossip(topic, peers)
             }
             lastPublished.entries.removeIf { (topic, lastPub) ->
-                (time - lastPub > fanoutTTL)
+                (curTime() - lastPub > coreParams.fanoutTTL.toMillis())
                     .whenTrue { fanout.remove(topic) }
             }
 
-            (mesh.keys.toSet() + fanout.keys).forEach { topic ->
-                val gossipPeers = (getTopicPeers(topic) - mesh[topic]!! - (fanout[topic] ?: emptyList()))
-                    .shuffled(random).take(DGossip)
-                submitGossip(topic, gossipPeers)
-            }
-
             mCache.shift()
+
             flushAllPending()
         } catch (t: Exception) {
             logger.warn("Exception in gossipsub heartbeat", t)
         }
     }
 
-    private fun prune(peer: PeerHandler, topic: String) = addPendingRpcPart(
-        peer,
-        Rpc.RPC.newBuilder().setControl(
-            Rpc.ControlMessage.newBuilder().addPrune(
-                Rpc.ControlPrune.newBuilder().setTopicID(topic)
-            )
-        ).build()
-    )
+    private fun emitGossip(topic: Topic, excludePeers: Collection<PeerHandler>) {
+        val ids = mCache.getMessageIds(topic)
+        if (ids.isEmpty()) return
+
+        val shuffledMessageIds = ids.shuffled(random).take(params.maxIHaveLength)
+        val peers = (getTopicPeers(topic) - excludePeers)
+            .filter { score.score(it) >= score.params.gossipThreshold && !isDirect(it) }
+
+        peers.shuffled(random)
+            .take(max((params.gossipFactor * peers.size).toInt(), coreParams.DLazy))
+            .forEach { enqueueIhave(it, shuffledMessageIds) }
+    }
 
-    private fun graft(peer: PeerHandler, topic: String) = addPendingRpcPart(
+    private fun graft(peer: PeerHandler, topic: Topic) {
+        mesh.getOrPut(topic) { mutableSetOf() }.add(peer)
+        enqueueGraft(peer, topic)
+        notifyMeshed(peer, topic)
+    }
+
+    private fun prune(peer: PeerHandler, topic: Topic) {
+        setBackOff(peer, topic)
+        enqueuePrune(peer, topic)
+        if (mesh[topic]?.remove(peer) == true) {
+            notifyPruned(peer, topic)
+        }
+    }
+
+    private fun iWant(peer: PeerHandler, messageIds: List<MessageId>) {
+        messageIds.forEach { iWantRequests[peer to it] = curTime() }
```

Comment:
It seems like the spec is suggesting to track only one random message-id, not all of the message-ids.
> For each IHAVE advertisement which elicits an IWANT request, the router tracks a random message ID within the advertised set.

Reply:
Right, that makes sense! Else the penalty can be too large for a single missed `IWANT`